### PR TITLE
add: env-exposer program and  reference test snap  

### PR DESCRIPTION
**Pull request by locnnil**
Add reference snap that uses `env-exporter` to export automatically environment variables

### CODE REVIEW

File: snap/snapcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+  envtester:
+    source: .
+    plugin: nil
+    build-snaps: [ go/latest/stable ]
+    override-build: |
+      GOOS=linux
+      [ "$CRAFT_ARCH_BUILD_FOR" == "arm64" ] && GOARCH=arm64 go build
+      [ "$CRAFT_ARCH_BUILD_FOR" == "amd64" ] && GOARCH=amd64 go build
+      [ "$CRAFT_ARCH_BUILD_FOR" == "riscv64" ] && GOARCH=riscv64 go build
```
Comment (farshidtz) [inline snap/snapcraft.yaml]:
```suggestion
      GOARCH=$CRAFT_ARCH_BUILD_FOR CGO_ENABLED=0 go build
```

File: snap/local/bin/env-exporter [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    key=$(echo "$key" | sed 's/"//g')
+    value=$(echo "$value" | sed 's/"//g')
+
+    hash_table["$key"]="$value"
+  done
+}
+
+check_num_at_start() {
+  if [[ "$1" =~ ^[0-9]+ ]]; then
+    debug "ERROR: Environment variable shouldn't begin with a number\n"
```
Comment (farshidtz) [inline snap/local/bin/env-exporter]:
```suggestion
    debug "ERROR: Environment variable name shouldn't begin with a number: $1\n"
```

File: snap/local/bin/env-exporter [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+export_vars() {
+  declare -n table=$1
+  local nk
+  for key in "${!table[@]}"; do
+
+    convert_keys $key nk
+    [ $? -ne 0 ] && continue
+    debug "old: $key -> new: $nk"
```
Comment (farshidtz) [inline snap/local/bin/env-exporter]: Why not print this in `convert_keys()`?

File: snap/local/bin/env-exporter [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+check_num_at_start() {
+  if [[ "$1" =~ ^[0-9]+ ]]; then
+    debug "ERROR: Environment variable shouldn't begin with a number\n"
+    return 1
+  fi
+  return 0
+}
+
+convert_keys() {
+  local fmt=$1
```
Comment (farshidtz) [inline snap/local/bin/env-exporter]:
```suggestion
  local key=$1
```
or
```suggestion
  local old=$1
```

File: snap/local/bin/env-exporter [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+    export "$nk=${table[$key]}"
+  done
+}
+
+handle_envs() {
+  json_str="$@"
+  [ -z "$json_str" ] && return
+
+  debug "[DEBUG] snapctl json:\n$json_str"
```
Comment (farshidtz) [inline snap/local/bin/env-exporter]:
```suggestion
  debug "snapctl json:\n$json_str"
```

File: snap/local/bin/env-exporter [hunk 6]
```diff
@@ -0,0 +1,129 @@
+#!/bin/bash
+
+debug() {
+  [ "$DEBUG" != "true" ] && return
+  echo -e "[DEBUG] $1"
```
Comment (farshidtz) [inline snap/local/bin/env-exporter:5]:
```suggestion
  echo -e "[DEBUG] $1\n"
```
Then remove the trailing line breaks from all the calls.

File: snap/local/bin/env-exporter [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+handle_envfile() {
+  local envfile=$1
+  [ -z "$envfile" ] && return
+
+  debug "Environment file path: $envfile"
+
+  if [ ! -f "$envfile" ]; then
+    debug "\nERROR: File not found: $envfile\n"
```
Comment (farshidtz) [inline snap/local/bin/env-exporter]: This is an error message that should appear regardless.

```suggestion
    echo -e "\nERROR: File not found: $envfile\n"
```

File: snap/local/bin/env-exporter [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    return 1
+  fi
+
+  if [ ! -r "$envfile" ]; then
+    debug "\nERROR: File not readable: $envfile\n"
```
Comment (farshidtz) [inline snap/local/bin/env-exporter]:
```suggestion
    echo -e "\nERROR: File not readable: $envfile\n"
```

**Review by farshidtz — COMMENTED** [review]
Looks great. Some minor suggestions.

File: env [hunk 9]
```diff
@@ -0,0 +1,7 @@
+
+ORACULAR=Oriole
+NOBLE=Numbat
+MANTIC=Minotaur
+LUNAR=Lobster
+KINETIC=Kudu
+JAMMY=Jellyfish
```
Comment (farshidtz) [inline env]:
```suggestion
# Ubuntu releases
ORACULAR=Oriole # latest and greatest!
NOBLE=Numbat
MANTIC=Minotaur
LUNAR=Lobster
KINETIC=Kudu
JAMMY=Jellyfish
```

File: README.md [hunk 10]
```diff
@@ -0,0 +1,27 @@
+# Envtester snap
+
+This snap is a PoC for the `env-exporter` program.
+
+`env-exporter` is a program that can dinamically load snap options into environment variables
```
Comment (farshidtz) [inline README.md]:
```suggestion
`env-exporter` is a program that can dynamically load snap options into environment variables
```

File: snap/snapcraft.yaml [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    build-for: [riscv64]
+
+grade: stable
+confinement: strict
+
+apps:
+  envtester:
+    plugs: [ home ]
+    command-chain: [ bin/env-exporter ]
+    command: envtester
```
Comment (farshidtz) [inline snap/snapcraft.yaml]: To test the global vs local:
- Change this to a oneshot service
- Replicate to have two services with different names
- Call them myapp1 and myapp2 for clarity?
- Restart the apps every time some new envs have been set

**Comment by farshidtz** [comment]
```
$ sudo snap set envtester env.good.bad=test # should not be accepted / converted
$ sudo snap run --shell envtester.myapp1
# env | grep GOOD
GOOD=bad:test
```

**Comment by locnnil** [comment]
@farshidtz In that case that you pointed, it's read as a global env!

**Comment by farshidtz** [comment]
Thanks. Merged into main without the last commit.

Superseded by #3

**Pull request closed without merging**